flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java

```diff
@@ -96,332 +222,274 @@ public void initializeState(StateInitializationContext context) throws Exception
-		private final TypeSerializer<OT> serializer;
-
-		private final Object checkpointLock;
-		private final SourceFunction.SourceContext<OT> readerContext;
-
-		private final Queue<TimestampedFileInputSplit> pendingSplits;
-
-		private TimestampedFileInputSplit currentSplit;
-
-		private volatile boolean isSplitOpen;
-
-		private SplitReader(FileInputFormat<OT> format,
-					TypeSerializer<OT> serializer,
-					SourceFunction.SourceContext<OT> readerContext,
-					Object checkpointLock,
-					List<TimestampedFileInputSplit> restoredState) {
-
-			this.format = checkNotNull(format, "Unspecified FileInputFormat.");
-			this.serializer = checkNotNull(serializer, "Unspecified Serializer.");
-			this.readerContext = checkNotNull(readerContext, "Unspecified Reader Context.");
-			this.checkpointLock = checkNotNull(checkpointLock, "Unspecified checkpoint lock.");
-
-			this.shouldClose = false;
-			this.isRunning = true;
+	private void loadSplit(TimestampedFileInputSplit split) throws IOException {
+		Preconditions.checkState(state != ReaderState.READING && state != ReaderState.CLOSED, "can't load split in state %s", state);
+		Preconditions.checkNotNull(split, "split is null");
+		LOG.debug("load split: {}", split);
+		currentSplit = split;
+		format.openInputFormat();
+		if (format instanceof CheckpointableInputFormat && currentSplit.getSplitState() != null) {
+			// recovering after a node failure with an input
+			// format that supports resetting the offset
+			((CheckpointableInputFormat<TimestampedFileInputSplit, Serializable>) format).
+					reopen(currentSplit, currentSplit.getSplitState());
+		} else {
+			// we either have a new split, or we recovered from a node
+			// failure but the input format does not support resetting the offset.
+			format.open(currentSplit);
+		}
 
-			this.pendingSplits = new PriorityQueue<>();
+		// reset the restored state to null for the next iteration
+		currentSplit.resetSplitState();
+	}
 
-			// this is the case where a task recovers from a previous failed attempt
-			if (restoredState != null) {
-				this.pendingSplits.addAll(restoredState);
-			}
+	private void switchState(ReaderState newState) {
+		if (state != newState) {
+			Preconditions.checkState(state.canSwitchTo(newState), "can't switch state from terminal state %s to %s", state, newState);
+			LOG.debug("switch state: {} -> {}", state, newState);
+			state = newState;
 		}
+	}
 
-		private void addSplit(TimestampedFileInputSplit split) {
-			checkNotNull(split, "Cannot insert a null value in the pending splits queue.");
-			synchronized (checkpointLock) {
-				this.pendingSplits.add(split);
-			}
-		}
+	@Override
+	public void processWatermark(Watermark mark) throws Exception {
+		// we do nothing because we emit our own watermarks if needed.
+	}
 
-		public boolean isRunning() {
-			return this.isRunning;
-		}
+	@Override
+	public void dispose() throws Exception {
+		super.dispose();
+		checkpointedState = null;
```

Comment:
On quite a few places `dispose` also `closes` resources. I have no idea about the actual contract though.
But it's probably safer to also call `format.close` etc from here.